contributor-docs/api.md

```diff
@@ -0,0 +1,194 @@
+# API Guidelines
+
+## Overview
+
+## React
+
+### Writing a component
+
+**Checklist**
+
+- [ ] Create entrypoint file
+- [ ] Create component file
+- [ ] Create test file
+- [ ] Create stories
+  - [ ] Create default stories with playground
+  - [ ] Create feature stories
+- [ ] Create e2e test
+  - [ ] Add support for VRT
+  - [ ] Add support for AAT
+
+```tsx
+type ExampleComponentProps = {}
+
+function ExampleComponent(props: ExampleComponent) {
+  // Setup, state
+  const [state, setState] = useState(defaultValue)
+  const value = 1
+
+  // Handlers
+
+  // Effects
+  useEffect(() => {
+    // ...
+  }, [])
+
+  // Output
+  return <ExampleComponentOutput state={state} />
+}
+```
+
+<table>
+<thead><tr><th>Unpreferred</th><th>Preferred</th></tr></thead>
+<tbody>
+<tr><td>
+
+```tsx
+const ExampleComponent: React.FC<ExampleProps = () => {
+    // ...
+};
+```
+
+</td><td>
+
+```tsx
+function ExampleComponent(props: ExampleComponentProps) {
+  // ...
+}
+```
+
+</td></tr>
+</tbody></table>
+
+#### When to use `React.forwardRef`
+
+- This might not be needed anymore with React 19?
+
+### Server-side rendering
+
+- Components should support SSR by default
+- Components should not rely on client-side state for layout (avoid layout
+  thrashing)
+- Components should progressively enhance as JavaScript loads on a page
+
+### Props
+
+#### Container props
+
+These are props that are provided consistently across components and are applied
+to the containing element rendered by the component.
+
+- `className`
+- `data-testid`
+- `React.ComponentPropsWithoutRef<...>`, typically seen as spread props: `{...rest}`
+
+The position of these props should be consistently applied to the same element
+across major versions. These should be tested.
+
+#### Controllable props
+
+- Works with and without props being passed
+- There should be a `value` and `onChange` or `onValueChange` prop when
+  controlled (naming doesn't have to be `value`)
+- Do we support mixing between modes?
+- `useControllableState` hook
+
+#### Naming boolean props
+
+- Boolean props could be tracking a state value or are opt-in to behavior or
+  appearance
+  - State behavior: `<ExampleComponent open={isOpen} />`
+  - Opt-in: `<ExampleComponent fullWidth />`
+- When naming "opt-in" variants, consider the default value and what it should
+  be. Name the prop according to what they will "opt-in" to.
+
+#### Naming variant props
+
+- Any specific naming conventions?
+- Should they always have a default value? (Maybe the one we expect to be most
+  used?)
+
+#### `useCallback` and `useMemo`
```

Comment:
This won't be really necessary after the React Compiler, so maybe  we can skip it